### PR TITLE
Motorway junc

### PR DESCRIPTION
**Pull request by ivanPyrohivskyi**

### CODE REVIEW

File: OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java [hunk 1]
```diff
@@ -341,13 +341,13 @@ public String getDestinationRef(boolean direction) {
 			}
 			//return names.get(region.refTypeRule);
 		}
-		return null;
+		return getRef(lang, transliterate, direction);
 	}
 
 	public String getDestinationName(String lang, boolean transliterate, boolean direction) {
-		//Issue #3289: Treat destination:ref like a destination, not like a ref
+		/*//Issue #3289: Treat destination:ref like a destination, not like a ref
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java:348]: code needs to be deleted and not commented

File: OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java [hunk 2]
```diff
@@ -15,11 +16,19 @@ public void setRef(String ref) {
         this.ref = ref;
     }
 
-    public String getExitStreetName() {
-        return exitStreetName;
+    public String getDestinationRef() {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java:19]: destination ref has nothing to do with exit info, so it shouldn't be stored here

File: OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java [hunk 3]
```diff
@@ -765,23 +765,23 @@ public boolean isExitPoint() {
...
-						if (pointNameTypes[i][j] == region.nameTypeRule) {
-							return point[j];
-						}
-					}
-				}
-			}
-		}
-		return null;
-	}
+//	public String getExitName() {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java:768]: no need to delete exit name might be useful in future

File: OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java [hunk 4]
```diff
@@ -15,11 +16,19 @@ public void setRef(String ref) {
...
+        return destinationRef;
     }
 
-    public void setExitStreetName(String exitStreetName) {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java:22]: no need to delete exit name might be useful in future

File: OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java [hunk 5]
```diff
@@ -15,11 +16,19 @@ public void setRef(String ref) {
...
-        this.exitStreetName = exitStreetName;
+    public void setDestinationRef(String destinationRef) {
+        this.destinationRef = destinationRef;
+    }
+
+    public String getDestinationName() {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java:27]: destination name has nothing to do with exit info (which is related only to motorway_exit), so it shouldn't be stored here.

Note: destination + destination ref should work on all points even not on motorway junction and it used to work

File: OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java [hunk 6]
```diff
@@ -395,50 +397,79 @@ public double getRouteVisibleAngle() {
...
-							ctx.getSettings().MAP_TRANSLITERATE_NAMES.get(), next.isForwardDirection()));
-					if (s.getObject().isExitPoint() && next.getObject().getHighway().equals("motorway_link")) {
+					info.setDestinationName(destinationName);
+
+					if (s.getObject().isExitPoint() && current.getObject().getHighway().equals("motorway_link")) {
 						ExitInfo exitInfo = new ExitInfo();
-						exitInfo.setRef(next.getObject().getExitRef());
-						exitInfo.setExitStreetName(next.getObject().getExitName());
+						exitInfo.setRef(current.getObject().getExitRef());
+						//exitInfo.setExitStreetName(next.getObject().getExitName());
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java:448]: exitInfo.setExitName(next.getObject().getExitName());

File: OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java [hunk 7]
```diff
@@ -395,50 +397,79 @@ public double getRouteVisibleAngle() {
 						// Consider roundabout end.
 						info.routeEndPointOffset = roundAboutEnd;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java:398]: This code needs to be moved to backend.
Suggested place - 
RouteSegmentResult

getTurnFollowDestinationRef(String lang, boolean transliterate)

getTurnFollowDestinationNames(String lang, boolean transliterate)

getTurnFollowStreetNames(String lang, boolean transliterate)

getTurnFollowRef(String lang, boolean transliterate)

File: OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java [hunk 8]
```diff
@@ -318,7 +318,7 @@ public String getRef(String lang, boolean transliterate, boolean direction) {
 		return null;
 	}
 
-	public String getDestinationRef(boolean direction) {
+	public String getDestinationRef(String lang, boolean transliterate, boolean direction) {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java:321]: Delete commented code above (as deed loop):
```
//if (getDestinationRef(direction) != null) {
		//	return getDestinationRef(direction);
		//}
		
```

Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java:318]: Fixed

File: OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java (hunk 1 above)
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java]: Ok

File: OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java (hunk 2 above)
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java]: Ok

File: OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java (hunk 3 above)
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java]: Ok

File: OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java (hunk 4 above)
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java]: Ok

File: OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java (hunk 5 above)
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/router/ExitInfo.java]: Ok

File: OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java (hunk 6 above)
Comment (ivanPyrohivskyi) [reply OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java]: Ok

File: OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java (hunk 8 above)
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/binary/RouteDataObject.java:318]: Ok

File: OsmAnd-java/src/main/java/net/osmand/router/RouteSegmentResult.java [hunk 9]
```diff
@@ -39,6 +39,11 @@
 	// this make not possible to make turns in between segment result for now
 	private TurnType turnType;
 
+	// Evaluates street name that the route follows after turn within specified distance.
+	// It is useful to find names for short segments on intersections
+	private static float distanceSeekStreetName = 150;
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteSegmentResult.java:44]: Constants guidelines + final as DIST_TO_SEEK_STREET_NAME, DIST_TO_SEEK_DEST

**Pull request merged**